x/smart/cache.go

```diff
@@ -0,0 +1,128 @@
+// Copyright 2023 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package smart
+
+import (
+	"context"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/Jigsaw-Code/outline-sdk/dns"
+	"golang.org/x/net/dns/dnsmessage"
+)
+
+// canonicalName returns the domain name in canonical form. A name in canonical
+// form is lowercase and fully qualified. Only US-ASCII letters are affected. See
+// Section 6.2 in RFC 4034.
+func canonicalName(s string) string {
+	return strings.Map(func(r rune) rune {
+		if r >= 'A' && r <= 'Z' {
+			r += 'a' - 'A'
+		}
+		return r
+	}, s)
+}
+
+type cacheEntry struct {
+	key    string
+	msg    *dnsmessage.Message
+	expire time.Time
+}
+
+// cacheResolver is a very simple caching [dns.Resolver].
+// It doesn't use the response TTL and doesn't cache empty answers.
+// It also doesn't dedup duplicate in-flight requests.
+type cacheResolver struct {
+	resolver dns.Resolver
+	cache    []cacheEntry
+	mux      sync.Mutex
+}
+
+var _ dns.Resolver = (*cacheResolver)(nil)
+
+func newCacheResolver(resolver dns.Resolver, numEntries int) dns.Resolver {
+	return &cacheResolver{resolver: resolver, cache: make([]cacheEntry, numEntries)}
+}
+
+func (r *cacheResolver) RemoveExpired() {
+	now := time.Now()
+	last := 0
+	r.mux.Lock()
+	defer r.mux.Unlock()
+	for _, entry := range r.cache {
+		if entry.expire.After(now) {
+			r.cache[last] = entry
+			last++
+		}
+	}
+	r.cache = r.cache[:last]
```

Comment:
*nit* any standard library would do the job? for example, [`slices.DeleteFunc`](https://pkg.go.dev/slices#DeleteFunc)?

```suggestion
	r.cache = slices.DeleteFunc(r.cache, func(entry cacheEntry) bool {
		return !entry.expire.After(now)
	})
```